More End of Chapter Exercises

alfordch opened the pull request:
Add mixed-up-code and multiple choice exercises to the following chapters:
- branching
- structs
- vectors in cpp

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+                }
+    =====
+            }
```
I mixed up these last three ending brackets and it won't let me get the answer right - I wonder if there's a way to mark which one is which? It's annoying to miss the problem because of these.

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    Construct a block of code that counts how many songs in ``album`` start with b. Put the necessary blocks of code in the correct order.
+    -----
+    #include <vector>
+    #include <string>
```
<vector> and <string> don't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    :adaptive:
+
+    Suppose ``album`` has already been defined as ``vector<string> album = {"imagine", "needy", "NASA", "bloodline", "fake smile", "bad idea", "make up", "ghostin", "in my head", "7 rings", "thank u, next", "break up with your girlfriend, i'm bored"}``
+    Construct a block of code that counts how many songs in ``album`` start with b. Put the necessary blocks of code in the correct order.
```
Maybe specify that album needs to be declared before count

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+    int main() {
+    =====
+        vector<double> temps = {82.0, 76.8, 74.3, 58.8, 79.2, 73.4, 80.1};
```
can't see <double>

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    Suppose you have the following two vectors to describe the weekly forecast ``vector<double> temps = {82.0, 76.8, 74.3, 58.8, 79.2, 73.4, 80.1}``,
+    ``vector<double> precip = {0.00, 0.30, 0.60, 0.90, 0.10, 0.20, 0.80}``. Your family will go to the beach if the temperature at least 75 degrees and the chance
+    of precipitation is less than 50%.  Construct a block of code that counts how many days your family can hit the beach on your vacation.
+    Put the necessary blocks of code in the correct order.
```
Maybe specify the order that the variables need to be declared

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    :adaptive:
+
+    Suppose you have the following vector ``vector<string> nouns = {"cereal", "Cocoa Puffs", "Mario", "luigi", "Aerosmith"}``.
+    Construct a block of code that creates a vector of the proper nouns in ``nouns`` called ``proper``.
```
Briefly define a proper noun, e.g., a noun that is capitalized

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    -----
+    #include <vector>
+    #include <string>
+    using namespace std;
```
Can't see <vector> and <string>

alfordch replied:
All of the "<" and ">" should now show, the formatting on my end was made incorrectly, sorry! I know that all of the questions on this parsons page had a similar problem.

laura-burdick replied:
It looks good now, thank you!

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<string> nouns = {"cereal", "Cocoa Puffs", "Mario", "luigi", "Aerosmith"};
+    =====
+        vector<string> proper = {};
```
can't see <string>

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<string> proper = {};
+    =====
+        for (int i = 0; i < nouns.size(); i++) {
```
++i not i++; size_t rather than int?

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose you have the following function ``howMany`` and vector ``exclamations``
```
Period after exclamations

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+
+        int howMany (const vector<string>& vec, char let) {
+            int count = 0;
+            for (int i = 0; i < vec.size(); i++) {
```
size_t?
++i rather than i++

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+        int howMany (const vector<string>& vec, char let) {
+            int count = 0;
+            for (int i = 0; i < vec.size(); i++) {
+                for (int k = 0; k < vec[i].size(); k++) {
```
size_t?
++k not k++

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+
+    -----
+    #include <vector>
+    #include <string>
```
<vector> and <string> don't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+    int main() {
+    =====
+        vector<string> exclamations = {"what?!", "how???", "fine!", "STOP.", "yay!!!!!", "ugh...!"};
```
<string> doesn't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<string> exclamations = {"what?!", "how???", "fine!", "STOP.", "yay!!!!!", "ugh...!"};
+    =====
+        vector<char> punc = {'.', '!', '?'};
```
<char> doesn't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<char> punc = {'.', '!', '?'};
+    =====
+        vector<int> counts = {};
```
<int> doesn't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<int> counts = {};
+    =====
+        for (int i = 0; i < punc.size(); i++) {
```
size_t?
++i not i++

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    Suppose you have the following function ``howMany`` and vector ``exclamations``
+    Construct a block of code that counts how many times ".", "!", and "?" occur in ``exclamations``.
+    Save the counts to a vector with "." count as the first element, "!" count as the second, and "?" count as the third.
+    Put the necessary blocks of code in the correct order.
```
Maybe specify order variables need to be declared

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    Write the function ``endsEven`` that takes a vector and removes elements from the end of the vector until
+    it ends with an even number. Put the necessary blocks of code in the correct order.
+    -----
+    void endsEven (vector<int> &vec) {
```
<int> doesn't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    -----
+    void endsEven (vector<int> &vec) {
+    =====
+    vector endsEven (vector<int> &vec) { #distractor
```
<int> doesn't show up

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        while (vec.back() % 2 != 0) {
+    =====
+        for (int i = 0; i < vec.size(); i++) { #paired
```
size_t?
++i not i++

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        for (int i = 0; i < vec.size(); i++) { #paired
+    =====
+        vec.pop_back();
```
This needs to be indented one more time

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    function should return a vector of ``num`` integers that are between 1 and ``max``, inclusive.
+    Put the necessary blocks of code in the correct order.
+    -----
+    vector<int> randomNums (int num, int max) {
```
can't see <int> (throughout problem)

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        vector<int> randomVec(num);
+    =====
+        for (int i = 0; i < num; i++) {
```
++i not i++ (throughout problem)

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        for (int i = 0; i < num; i++) {
+    =====
+        for (int i = 0; i <= randomVec.size(); i++) { #paired
```
size_t?

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    You can assume the size of the vector will always be even.
+    Put the necessary blocks of code in the correct order.
+    -----
+    void weird_print (vector<int> vec) {
```
can't see <int>

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+
+    Write the function ``randomNums`` that takes two integers: ``num`` which is the number of random numbers
+    you wish to generate, and ``max``, which is the maximum value of random number you wish to generate.  Your
+    function should return a vector of ``num`` integers that are between 1 and ``max``, inclusive.
```
Because we teach students to always pass vectors by reference or const reference, I think it would be best to re-write this function so it took randomNums as a parameter by reference and then filled it with random numbers

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        int half = vec.size() / 2;
+    =====
+        for (int i = vec.size() - 1; i >= half; i--){
```
size_t?
--i not i--

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        }
+    =====
+        for (int h = 0; h < half; h++) {
```
++h not h++

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        for (int i = vec.size() - 1; i >= half; i--){
+    =====
+            cout << vec[i-half] << ' ';
```
spacing: vec[i - half]

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+        }
+    =====
+        cout << endl;
+    =====
```
I think this is not a very intuitive way to solve this problem. My first thought is to iterate from the middle of the vector to the beginning, printing out numbers. Then, in a second loop, iterate from the middle of the vector to the end. I think that the current solution is not the easiest way to solve this, so I would either remove this question or simplify this.

laura-burdick reviewed (changes requested):
Thanks, Charles, for all of your work on this! I've left some comments - let me know if you have any questions.

alfordch commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,244 @@
+    =====
+    string printAddress (Employee& e) {  #paired 
+    =====
+        cout << e.name << " (" << e.id << ") lives at ";
```
That line was accidentally removed. Thank you for noticing that!

alfordch commented on _sources/vectors_in_cpp/Exercises.rst:
```diff
@@ -0,0 +1,455 @@
+
+.. mchoice:: vectors_in_cpp_mc4
+
+    Select all of the following statments that correctly make a copy of ``lauren``.
```
I changed some of the names, let me know if this reads any better.

laura-burdick replied:
this looks good!

alfordch commented:
> Thanks, Charles, for all of your work on this! I've left some comments - let me know if you have any questions.

I've gone ahead and worked through those comments. For the branching chapter, I removed those end of chapter files for me to completely redo, as I neglected to check the sequencing of the content before it, sorry! The parsonsprobs shouldn't be recognizing the "}" lines as different, so the problems should not specify which one comes first. I realize that would make these problems pretty frustrating. The issue there should have been caused by bad formatting in lines before, so if I have fixed those correctly, that issue should be gone. Let me know if that is still happening. Thank you!

laura-burdick commented on _sources/structs/Exercises.rst:
```diff
@@ -0,0 +1,229 @@
+        int mass;
+      };
+
+      int calculateDensity (Cube &c) {
```
c should be passed by const reference since c is not modified in any way

laura-burdick commented on _sources/structs/Exercises.rst:
```diff
@@ -0,0 +1,229 @@
+
+    - 0
+
+      + Since we pass a ``Student`` object by value to ``pourCoffee``, the function makes a copy of the object and does not modify the original. If you wanted the original value to change, pass it by reference!
```
The answer to this question changes because you changed the function from pass by value to pass by reference - this is an outdated answer

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    :numbered: left
+    :adaptive:
+
+    Let's write the code for the ``convertToHumanAge`` function. ``convertToHumanAge``
```
This one is doing that weird thing where I can see the question, but the Parson's problem doesn't have any code blocks in it. I'm assuming you were trying to delete this question?

alfordch replied:
All of the questions that you noted were not showing up correctly in the structs chapter and vectors chapter now have updated ID's, they should now display on the page with all of the correct blocks along with the question text.

laura-burdick replied:
Thanks! They show up now.

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    following format: ``name (id) lives at houseNumber in state, postalAddress``.
+    Put the necessary blocks of code in the correct order.
+    -----
+    void printAddress (Employee &e) {
```
e should be passed by const reference because it's not changed in the function

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    :numbered: left
+    :adaptive:
+
+    Sometimes employees will move around and thus we'll need to update their addresses.
```
This parson's problem also looks blank

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    :numbered: left
+    :adaptive:
+
+    Let's write the code for the ``storeEmployeeData`` function. ``storeEmployeeData`` should be a void function that takes an empty ``Employee`` as an input.
```
This one also looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,292 @@
+    =====
+        }
+    =====
+    }
```
This is fixed!

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose you have the vector ``vector<string> words = {"car", "cat", "switch", "princess"}``
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose ``album`` has already been defined as ``vector<string> album = {"Welcome", "Meals to Dinner", "Truly Unique", "Live At Home", "Theme Music", "Soul", "This Is Me", "Jamboree", "Dear You"}``
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose you have the following two vectors to describe the weekly forecast ``vector<double> temps = {82.0, 76.8, 74.3, 58.8, 79.2, 73.4, 80.1}``,
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose you have the following vector ``vector<string> nouns = {"cereal", "Cocoa Puffs", "Mario", "luigi", "Aerosmith"}``.
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Suppose you have the following function ``howMany`` and vector ``exclamations``.
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    :numbered: left
+    :adaptive:
+
+    Write the function ``endsEven`` that takes a vector and removes elements from the end of the vector until
```
This parson's problem looks blank

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    =====
+        }
+    =====
+        return randomVec #distractor
```
semi colon at end?

laura-burdick commented:
Thanks for making these changes! I think the issue where the ending bracket lines have to go in a certain order is fixed.

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    four times its ``age`` in unicorn years plus twice the ``hornLength``.
+    Put the necessary blocks of code in the correct order.
+    -----
+    int convertToHumanAge (Unicorn &u) {
```
u should be passed by const reference because the function doesn't modify it

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    If a unicorn ``isSparkly``, then its equivalent human age is three times its ``age`` in unicorn years
+    plus the ``hornLength``. If a unicorn is not sparkly, then its equivalent human age is
+    four times its ``age`` in unicorn years plus twice the ``hornLength``.
+    Put the necessary blocks of code in the correct order.
```
Specify in in the instructions that you must use an else statement (this could be solved without using the else statement, but then you get it wrong)

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    :adaptive:
+
+    Sometimes employees will move around and thus we'll need to update their addresses.
+    Let's write the code for the ``updateAddress`` function. updateAddress takes an
```
updateAddress at beginning of sentence should be in code font

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    ``Employee`` and a new ``Address`` as parameters and sets the employee's address to the new address.
+    Put the necessary blocks of code in the correct order.
+    -----
+    void updateAddress (Employee &e, Address &a) {
```
a should be passed by const ref

laura-burdick commented on _sources/structs/mixed_up_code.rst:
```diff
@@ -0,0 +1,236 @@
+    =====
+        cin >> e.address.postalAddress;
+    =====
+        return Employee e;  #paired
```
Why is this one paired with the previous line? It seems like maybe it shouldn't be paired

alfordch replied:
Yes, this should have been a #distractor, thanks. Fixed now!

laura-burdick commented on _sources/vectors_in_cpp/mixed_up_code.rst:
```diff
@@ -0,0 +1,259 @@
+    =====
+        int count = 0;
+    =====
+        for (size_t i = 0; i < 7; ++i) {
```
this can be int not size_t

laura-burdick reviewed (changes requested):
Just a few more small comments, but looking good! Thanks!

laura-burdick commented:
Thanks, Charles!
